Integrate nrfx 2.1.0

anangl opened the pull request:
Import relevant files from nrfx 2.1.0 and provide a glue layer that adapts nrfx to be usable in TF-M as a host environment.

For the beginning, two nrfx drivers are imported: nrfx_nvmc and nrfx_uarte, as these are likely to be used in implementations of CMSIS Driver_Flash and Driver_USART. We'll add more drivers when it becomes necessary. From nrfx HAL, files for all peripherals available in nRF5340 application core or nRF9160 are imported.

anangl committed:
platform: nordic_nrf: Import relevant files from nrfx 2.1.0

nrfx is a standalone set of drivers for peripherals present in Nordic
Semiconductor's SoCs. For the user's convenience, the drivers come
with the MDK package. This package contains definitions of register
structures and bitfields for all supported SoCs, as well as startup
and initialization files for them.

Origin: nrfx
License: BSD 3-Clause
URL: https://github.com/NordicSemiconductor/nrfx/tree/v2.1.0
commit: 281cc2e178fd9a470d844b3afdea9eb322a0b0e8
Purpose: Provide peripheral drivers for Nordic SoCs
Maintained-by: External

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
platform: nordic_nrf: Add initial adaptation layer for nrfx

This commit adds a glue layer that adapts nrfx to be usable in TF-M
as a host environment. It provides also the cmsis.h header file that
redirects to CMSIS definitions brought by MDK, a part of nrfx.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
platform: nordic_nrf: Add initial cmake lists for nrf5340 and nrf9160

These lists will gather all files specific for those SoCs.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/nrfx_glue.c:
```diff
@@ -0,0 +1,49 @@
+
+void nrfx_critical_section_exit(void)
+{
+    --in_critical_section;
```
Should there be an assert for underflow here?
```suggestion
    NRFX_ASSERT(in_critical_section > 0);
    --in_critical_section;
```

anangl replied:
Applied.

ioannisg reviewed (commented):
@anangl quick comment, regarding the cmake files: we've agreed (at least for the initial tf-m porting phase) to use the full target names in the cmake targets for nrf platforms, so probably we need to name the initial .cmake files accordingly

anangl commented:
> so probably we need to name the initial .cmake files accordingly

@ioannisg For sure, if we decide to move the files I added to `platform/ext` (they are now added to `platform/ext/target/nordic_nrf/nrf5340/` and `platform/ext/target/nordic_nrf/nrf9160/`, and I named them after those directories). But I thought that we could instead include them from `platform/ext/nrf5340pdk_nrf5340_cpuapp.cmake` and `platform/ext/nrf9160dk_nrf9160.cmake`, respectively, which are to be added yet. This way those files could be reused by other boards based on nRF SoCs.

ioannisg commented:
> > so probably we need to name the initial .cmake files accordingly
> 
> @ioannisg For sure, if we decide to move the files I added to `platform/ext` (they are now added to `platform/ext/target/nordic_nrf/nrf5340/` and `platform/ext/target/nordic_nrf/nrf9160/`, and I named them after those directories). But I thought that we could instead include them from `platform/ext/nrf5340pdk_nrf5340_cpuapp.cmake` and `platform/ext/nrf9160dk_nrf9160.cmake`, respectively, which are to be added yet. This way those files could be reused by other boards based on nRF SoCs.

Oh, sure, I notice now they are not the root .cmake files, but they are added one level deeper

ioannisg commented on platform/ext/target/nordic_nrf/common/nrfx_config.h:
```diff
@@ -0,0 +1,57 @@
+ * without any suffixes, the following macro is provided that will translate
+ * their names according to the kind of the target that is built.
+ */
+#if defined(NRF_TRUSTZONE_NONSECURE)
```
That's probably OK. I wonder only if we are supposed to build non-secure images i the tf-m project :)

ioannisg reviewed (commented):
@anangl i don't have any concerns with this PR, other than whether we need to bring in the whole nrfx drivers suite, or just the required parts of it.

Btw, probably you can skip importing the nrf51_erratas and nrf52_erratas headers, unless i am missing something here.

anangl commented:
> @anangl i don't have any concerns with this PR, other than whether we need to bring in the whole nrfx drivers suite, or just the required parts of it.

@ioannisg Only two drivers are imported now: nrfx_nvmc and nrfx_uarte. But as I wasn't sure which peripherals we might need to access, I included HALs for all available in nRF5340 or nRF9160. I could probably limit this list even more now, as it's rather unlikely that we'd ever need here things like I2S, PWM, or QDEC. I'll update this PR trying to remove more unnecessary stuff.

> Btw, probably you can skip importing the nrf51_erratas and nrf52_erratas headers, unless i am missing something here.

I can't. Those are [included](https://github.com/NordicPlayground/trusted-firmware-m/pull/8/files#diff-6e362dd54e03c3dcc3517b838dac429aR40) unconditionally by `nrf_erratas.h`.

ioannisg commented:
> I can't. Those are included unconditionally by nrf_erratas.h.

Oh, I see. But that should probably changed in the upstream project, no?

anangl commented:
> > I can't. Those are included unconditionally by nrf_erratas.h.
> 
> Oh, I see. But that should probably changed in the upstream project, no?

Yes, in MDK, which is imported into nrfx.

ioannisg commented on platform/ext/target/nordic_nrf/common/nrfx/mdk/nrf5340_application.h:
```diff
@@ -0,0 +1,3462 @@
+/*
```
@anangl I've got a question about the vendor headers, e.g. this one:
In TF-M there seems to be a convention for the IRQ enum to be defined in a header called "platform_irq.h"
How is it going to work with this nrfx header?

anangl replied:
These headers (`platform_irq.h`) are only included by other vendor files, like `tfm_peripherals_def.h`, so IMO we don't need to provide such header, we can just include `nrf.h` instead.

ioannisg replied:
Yeah, or have a "dummy" platform_irq.h header that includes nrfx.h.
I'd say that's safer option - those headers might be there for historical reasons

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/nrfx_log.h:
```diff
@@ -0,0 +1,135 @@
+ * @param format printf-style format string, optionally followed by arguments
+ *               to be formatted and inserted in the resulting string.
+ */
+#define NRFX_LOG_ERROR(format, ...)
```
Will this eventually use `LOG_MSG` from tfm_log.h?

anangl replied:
Yes, I plan to add it later when I will be able to test it in action.

ioannisg commented on platform/ext/target/nordic_nrf/common/cmsis.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef __NORDIC_NRF_CMSIS_H__
+#define __NORDIC_NRF_CMSIS_H__
+
+#include <nrf.h>
```
@anangl in most portings, this header is part of 
`cmsis_core/Include` or `Device/Include` directories, and it includes
- `platform_regs.h`
- `platform_base_address.h`
- `platform_irq.h`

Although there does not seem to be any strict convention, I wonder how our cmsis.h could look like the other ones in the tree. your thoughts?

ioannisg reviewed (approved):
I suppose this integration is OK for now - might need to revisit things at a later stage

tejlmand reviewed (approved):
lgtm.

Just a question:
> But I thought that we could instead include them from platform/ext/nrf5340pdk_nrf5340_cpuapp.cmake and platform/ext/nrf9160dk_nrf9160.cmake, respectively, which are to be added yet.

Do you expect to do this in a followup PR ?

ioannisg commented:
> lgtm.
> 
> Just a question:
> 
> > But I thought that we could instead include them from platform/ext/nrf5340pdk_nrf5340_cpuapp.cmake and platform/ext/nrf9160dk_nrf9160.cmake, respectively, which are to be added yet.
> 
> Do you expect to do this in a followup PR ?

There is an open task for this - yes.
